follower/consensus_follower.go

```diff
@@ -0,0 +1,108 @@
+package consensus_follower
+
+import (
+	"context"
+
+	"github.com/reactivex/rxgo/v2"
+
+	"github.com/onflow/flow-go/cmd"
+	access "github.com/onflow/flow-go/cmd/access"
+	"github.com/onflow/flow-go/consensus/hotstuff/notifications/pubsub"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+func finalizedBlockProducer(anb *access.UnstakedAccessNodeBuilder) rxgo.Producer {
+	return func(ctx context.Context, next chan<- rxgo.Item) {
+		anb.FinalizationDistributor.AddOnBlockFinalizedConsumer(onBlockFinalizedConsumer(next))
+
+		select {
+		case <-anb.Ready():
+			// log: component startup complete
+		case <-ctx.Done():
+			// log: component startup aborted
+			return
+		}
+
+		<-ctx.Done()
+
+		// log: shutting down
+
+		<-anb.Done()
+	}
+}
+
+func onBlockFinalizedConsumer(next chan<- rxgo.Item) pubsub.OnBlockFinalizedConsumer {
+	return func(finalizedBlockID flow.Identifier) {
+		next <- rxgo.Of(finalizedBlockID)
+	}
+}
+
+func buildAccessNode(accessNodeOptions []access.Option) *access.UnstakedAccessNodeBuilder {
+	anb := access.FlowAccessNode(accessNodeOptions...)
+	nodeBuilder := access.NewUnstakedAccessNodeBuilder(anb)
+
+	nodeBuilder.BuildConsensusFollower()
+
+	return nodeBuilder
+}
+
+type Option func(c *Config)
+
+func getAccessNodeOptions(config *Config) []access.Option {
+	return []access.Option{
+		access.WithUpstreamAccessNodeID(config.upstreamAccessNodeID),
+		access.WithUnstakedNetworkBindAddr(config.bindAddr),
+		access.WithBaseOptions(getBaseOptions(config)),
+	}
+}
+
+func getBaseOptions(config *Config) []cmd.Option {
+	options := []cmd.Option{cmd.WithNodeID(config.nodeID)}
+	if config.bootstrapDir != "" {
+		options = append(options, cmd.WithBootstrapDir(config.bootstrapDir))
+	}
+	if config.dataDir != "" {
+		options = append(options, cmd.WithDataDir(config.dataDir))
+	}
+
+	return options
+}
+
+func GetFinalizedBlockProducer(
+	nodeID flow.Identifier,
+	upstreamAccessNodeID flow.Identifier,
+	bindAddr string,
+	opts ...Option,
+) rxgo.Producer {
```

Comment:
is `rxgo.Producer` a popular library? I would have preferred to just use the basic GoLang libarary to define our API. What extra does the `rxgo.Producer` gives us over simply returning a `read-only channel` ?

Reply:
the `rxgo` also tends to hide what is being produced (in this case `blocks`).

Reply:
I think RxGo is a pretty popular library for the observable pattern, in fact it's the official ReactiveX API for the go language: https://github.com/ReactiveX/RxGo

We definitely don't have to use it here, but I think it does a lot of the work for us, and it is very flexible from the Observer's perspective.

I only have to define the Producer function, and anyone who wants to create an Observable from it can do so while specifying various different options: https://github.com/ReactiveX/RxGo/blob/master/doc/options.md

They can create an observable that is lazy, uses a buffered channel, has a given context, [automatically does fanout](https://github.com/ReactiveX/RxGo/blob/master/README.md#connectable-observable) to multiple observers, etc.

Reply:
Can we bring this choice to a team discussion?

I was quite into functional programming and functional reactive programming a few years ago. And did try RxJS and other similar libraries, the experience wasn't good to be honest. For instance, I messed up the concept of hot/cold observables and made bug. The error handling was confusing especially when multiple observables are merged. Wasn't easy to debug when the consumer didn't receive an expected event. 

It does save some work, but also introduces some patterns and behaviors that not most golang developer would be familiar with. 

For instance, the "lazy and buffered" isn't clear to me what would happen if the buffered channel gets full, would the new message drop or blocking the producer? And what's beneficial being "lazy"? Does it come with a performance cost?

As this PR only introduces the Observable interfaces, it's not clear what's the full picture of how it'll be used by unstaked follower engines. What is "fanout to the multiple observers" use case? Is each observer a unstaked peer? If so, then would new staked peer connect connect and disconnect affect each other's consumption of the message? 

The original RX idea is from RxJava, and Java has a good support for generic to reuse those functions and maximize the benefit. However, in golang, with Observables, in order to provide reusable functions (like Map, ForEach), each item/message would have to be defined as `interface{}`. That means you lose type check from the compiler. For instance, we create an Observable of a stream of finalized block IDs, but the observer might cast it into block headers, then this buggy program would still compile. And it might not be easy to find out by tests, because the unittest for mismatching observable and observers could all pass. We would have to rely on integration tests to discover such mistakes.

I'm open to any choice, but just to bring up some "invisible cost" here.

Reply:
Thanks Leo! I think after some discussion with @huitseeker and @vishalchangrani, we agree this probably isn't the right time to introduce this library.

However, I would love to bring this to a team discussion at some point later on to explore the idea of having a standard way (e.g perhaps RxGo library, but perhaps something else) of implementing the observable pattern.

We currently have many examples throughout the codebase where the observable pattern is being used, but differ in various ways (e.g do we use channels or callbacks? are callbacks blocking? do they arrive in order?) and I think it could perhaps be valuable to have a standard way of dealing with this pattern.

Reply:
Thanks @zhangchiqing ! I think after some discussion with @huitseeker and @vishalchangrani, we agree this probably isn't the right time to introduce this library.

However, I would love to bring this to a team discussion at some point later on to explore the idea of having a standard way (e.g perhaps RxGo library, but perhaps something else) of implementing the observable pattern.

We currently have many examples throughout the codebase where the observable pattern is being used, but differ in various ways (e.g do we use channels or callbacks? are callbacks blocking? do they arrive in order?) and I think it could perhaps be valuable to have a standard way of dealing with this pattern.